Optimize enemy loading

WilliamFr0g opened the pull request:
Probably still kinda bad, but it's a lot less bad

I only just quickly swapped some calls for this to work in love 11 because it was built in love 10, it has not been formally tested

WilliamFr0g committed:
Optimize enemy loading

Fewer redundant calls to check if files exist (definitely), less string manipulation (I think), more efficient loading order (probably)

WilliamFr0g committed:
Fix uncommon daily challenge bug

Also removes mysterious parsing for a nonexistent "fireenemy" enemy property

WilliamFr0g committed:
PNG support

For all 1 mappacks that use .PNG instead of .png (looking at you, Aidan)

alesan99 commented on enemies.lua:
```diff
@@ -104,177 +128,269 @@ function enemies_load()
 		end
+	end
 
-		if enemiesdata[s].fireenemy then
```
"fireneemy" was the old name for "makesMarioShoot". This code was non-functional but it should be kept and fixed to say makesMarioShoot so there's case insensitivity in custom powerup properties.

WilliamFr0g replied:
Adding case insensitivity when it hasn't existed for years might do more harm than good at this point, last time this happened (with supersizeproperties) it broke some of my creations that used it and I had to lobby for an additional property to turn it back

alesan99 replied:
I see what you're saying, I don't like it but yeah backwards compatibility is more important at this point.

alesan99 reviewed (dismissed):
Haven't tested yet but some case-sensitivity code should be fixed and re-added

qixils committed:
Merge branch 'master' into pr/WilliamFr0g/591

qixils committed:
chore: reduce diff

qixils reviewed (approved):
Given how many times enemies are (re)loaded I definitely think we should have a perpetual in-memory cache (along the lines of a table mapping filepaths to enemies) somewhere to *really* optimize enemy loading, or at least smarter logic for when to clear the customenemies table, as currently enemies are still loaded on mappack change as well as reloaded on level change. So while this PR certainly cleans up the code and resolves issues, it doesn't feel especially faster (although I haven't benchmarked it, I see the code for it and trust that it was done). However I think it would make sense to save that for a future PR, as managing the cache alongside the base system seems nontrivial.

Tested briefly and everything seems functional to me.

EDIT: Did do a quick benchmark; I get about 0.3s (x2=0.6s) to load the Kirby mappack's enemies on this branch, and 0.4s (x2=0.8s) on the master branch. 33% improvement isn't too bad, but still a total of 0.6s being spent to actually get into a level in the mappack, with half of that being a useless reload, so definitely room to improve in the future.

qixils committed:
Merge branch 'master' into pr/WilliamFr0g/591

WilliamFr0g committed:
Fix rare bugs

Some very strangely-named enemies would trip up the parsing

WilliamFr0g committed:
Fix rare bug

You were able to load images with suffixes besides -icon and -tooltip and they would become the enemy's main image, which almost never was a problem but I did see this happen once

WilliamFr0g committed:
Fix Casing Issue

Not like you should have uppercase in enemy names anyway...
